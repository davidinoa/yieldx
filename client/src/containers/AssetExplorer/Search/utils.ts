interface InGridFilter {
  filter: string | number;
  filterTo: number;
  filterType: string;
  type: string;
  dateFrom: string;
  dateTo: string;
  values: string[];
}

export interface AgGridFilters {
  [key: string]: any;
}

const getFilter = (filter: InGridFilter) => {
  /**
   *  The filter object from Ag Grid API. In the case of filtering `name` column.
   *  { filter: "Asset Name", filterType: "text", type: "contains"}
   */
  let searchFilter = {};
  const {
    filter: filterContent,
    filterType,
    type,
    filterTo,
    values,
    dateFrom,
    dateTo
  } = filter;
  if (filterType === "text") {
    switch (type) {
      case "contains":
        searchFilter = { containsText: filterContent };
        break;
      case "notContains":
        searchFilter = { doesNotContainText: filterContent };
        break;
      case "equals":
        searchFilter = { equalsText: filterContent };
        break;
      case "notEqual":
        searchFilter = { doesNotEqualText: filterContent };
        break;
      case "startsWith":
        searchFilter = { startsWithText: filterContent };
        break;
      case "endsWith":
        searchFilter = { endsWith: filterContent };
        break;
      default:
        searchFilter = { containsText: filterContent };
        break;
    }
  } else if (filterType === "number") {
    switch (type) {
      case "equals":
        searchFilter = { equalsNumber: filterContent };
        break;
      case "notEqual":
        searchFilter = { doesNotEqualNumber: filterContent };
        break;
      case "lessThan":
        searchFilter = { isLessThanNumber: filterContent };
        break;
      case "lessThanOrEqual":
        searchFilter = { isLessThanOrEqualToNumber: filterContent };
        break;
      case "greaterThan":
        searchFilter = { isGreaterThanNumber: filterContent };
        break;
      case "greaterThanOrEqual":
        searchFilter = { isGreaterThanOrEqualToNumber: filterContent };
        break;
      case "inRange":
        searchFilter = {
          isGreaterThanOrEqualToNumber: filterContent,
          isLessThanOrEqualToNumber: filterTo
        };
        break;
      default:
        searchFilter = { equalsNumber: filterContent };
        break;
    }
  } else if (filterType === "date") {
    const convertedDateFrom = new Date(dateFrom);
    const convertedDateTo = new Date(dateTo);
    switch (type) {
      case "equals":
        searchFilter = { equalsDate: convertedDateFrom };
        break;
      case "greaterThan":
        searchFilter = { isGreaterThanDate: convertedDateFrom };
        break;
      case "lessThan":
        searchFilter = { isLessThanDate: convertedDateFrom };
        break;
      case "notEqual":
        searchFilter = { doesNotEqualDate: convertedDateFrom };
        break;
      case "inRange":
        searchFilter = {
          isGreaterThanOrEqualToDate: convertedDateFrom,
          isLessThanOrEqualToDate: convertedDateTo
        };
        break;
      default:
        searchFilter = { equalsDate: convertedDateFrom };
        break;
    }
  } else if (filterType === "set" && values) {
    // We may need to add `set` type Ag grid column filter params like cusip or isin
    // in the search request in the future.
    // searchFilter = { inText: values };
  }
  return searchFilter;
};

export const convertInGridFilters = (filters: AgGridFilters) => {
  const convertedFilters: { [index: string]: any } = {};
  Object.keys(filters).forEach(key => {
    const inGridFilter = filters[key];
    if (key === "minimumQuantity") key = "bond.minimumDenomination";
    convertedFilters[key] = getFilter(inGridFilter);
  });

  return convertedFilters;
};
