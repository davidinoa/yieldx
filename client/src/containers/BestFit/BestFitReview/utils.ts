import { getRatingValue } from "../../../util/ratings";
import YieldXPortfolio, {
  YieldXPosition
} from "../../../models/YieldXPortfolio";

export function getBondRatingsData(
  portfolio: YieldXPortfolio
): { name: string; value: number }[] {
  if (portfolio.getBondSleeve) {
    const ratingMarketValues: { [key: string]: number } = {};
    const bondPortfolio = portfolio.getBondSleeve();
    bondPortfolio.positions.forEach((position: YieldXPosition) => {
      const { asset, weight } = position;
      const rating = asset.rating ?? "NR";
      ratingMarketValues[rating] = (ratingMarketValues[rating] || 0) + weight;
    });
    return Object.entries(ratingMarketValues)
      .map(([rating, weight]) => {
        return {
          name: rating,
          value: weight
        };
      })
      .sort((a, b) => {
        return getRatingValue(a.name) - getRatingValue(b.name);
      });
  }
  return [];
}
