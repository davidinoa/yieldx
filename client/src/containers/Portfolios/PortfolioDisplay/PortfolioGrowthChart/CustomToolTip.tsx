import React from "react";
import styled from "styled-components";
import Decimal from "../../../../util/Decimal";

const Root = styled.div`
  background: var(--page);
  border: 1px solid var(--border-new);
  border-radius: 5px;
  width: 158px;
  height: 100px;
  padding: 0 7px;
  position: relative;
`;

const Value = styled.div`
  > span {
    font-family: Roboto Mono;
    font-size: 16px;
    line-height: 19px;
  }
  display: flex;
  justify-content: center;
  align-items: center;
  color: #ffffff;
  height: 70px;
`;

const Label = styled.div`
  border-top: 1px solid var(--border-new);
  width: 100%;
  > span {
    font-weight: bold;
    font-size: 10px;
    line-height: 12px;
    letter-spacing: 0.2em;
    text-transform: uppercase;
  }
  display: flex;
  justify-content: center;
  align-items: center;
  height: 30px;
`;

const Arrow = styled.div`
  width: 70px;
  height: 16px;
  overflow: hidden;
  position: absolute;
  right: 50%;
  transform: translateX(50%);
  bottom: -16px;
  :after {
    content: "";
    position: absolute;
    top: -20px;
    right: 48%;
    width: 20px;
    height: 20px;
    transform: rotate(45deg) translateX(50%);
    background: var(--page);
    border: 1px solid var(--border-new);
    border-radius: 2px;
  }
`;

type Props = {
  title?: string;
  value: string;
};

export default function CustomToolTip({ title, value }: Props) {
  return (
    <Root>
      <Value>
        <span>{`$${Decimal.format(parseFloat(value))}`}</span>
      </Value>
      <Label>
        <span>{!title ? "now" : title}</span>
      </Label>
      <Arrow />
    </Root>
  );
}
