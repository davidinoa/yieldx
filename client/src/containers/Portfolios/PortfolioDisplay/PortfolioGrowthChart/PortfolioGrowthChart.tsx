import React, { MutableRefObject, useEffect, useRef } from "react";
import ReactDOM from "react-dom";
import { Col } from "react-bootstrap";
import Chart from "chart.js";
import styled from "styled-components";
import PortfolioMetricCard from "components/PortfolioMetricCard/PortfolioMetricCard";
import { WealthProjection } from "providers/graphql/hooks";
import CustomToolTip from "./CustomToolTip";

interface Props {
  nav: number;
  wealthProjections: WealthProjection[];
  tooltipContent?: string;
}

const ChartCard = styled(Col)`
  padding: 1.5rem 0.5rem 0;
`;

const StyledRootCard = styled(PortfolioMetricCard)`
  margin-bottom: 3rem;
`;

function formatNumber(val: number): string {
  if (val < 1_000_000) return `$${Math.round(val / 1_000).toString()}K`;
  return `$${(val / 1_000_000).toFixed(2)}MM`;
}

const PortfolioGrowthChart = ({
  wealthProjections,
  nav,
  tooltipContent
}: Props) => {
  const chartRef: MutableRefObject<HTMLCanvasElement | null> = useRef(null);
  const [, setChart] = React.useState<null | Chart>(null);
  useEffect(() => {
    if (chartRef.current) {
      const canvas: HTMLCanvasElement = chartRef.current;
      const data = [
        {
          nav,
          lowerBar: nav,
          upperBar: nav
        },
        ...wealthProjections.map(({ percentiles }, i) => {
          if (i === 0 || i === 3 || i === 6) return undefined;
          return {
            nav: percentiles[1]?.nav,
            lowerBar: percentiles[0]?.nav,
            upperBar: percentiles[2]?.nav
          };
        })
      ].filter(cv => cv !== undefined) as {
        nav: number;
        lowerBar: number;
        upperBar: number;
      }[];

      const { max, min } = data.reduce(
        (acc, it) => {
          return {
            min: Math.min(acc.min, it.lowerBar),
            max: Math.max(acc.max, it.upperBar)
          };
        },
        {
          max: -Infinity,
          min: Infinity
        }
      );

      let stepSize = Math.ceil(max - nav) / 7;
      if (max - nav <= 140) stepSize = 10;
      else if (max - nav <= 1_400) stepSize = 100;
      const lineChart = new Chart(canvas, {
        type: "line",
        options: {
          tooltips: {
            // Disable the on-canvas tooltip
            enabled: false,

            custom(tooltipModel) {
              // Tooltip Element
              let tooltipEl = document.getElementById("chartjs-tooltip");

              // Create element on first render
              if (!tooltipEl) {
                tooltipEl = document.createElement("div");
                tooltipEl.id = "chartjs-tooltip";
                document.body.appendChild(tooltipEl);
              }

              // Hide if no tooltip
              if (tooltipModel.opacity === 0) {
                tooltipEl.style.opacity = "0";
                return;
              }

              // Set caret Position
              tooltipEl.classList.remove("above", "below", "no-transform");
              if (tooltipModel.yAlign) {
                tooltipEl.classList.add(tooltipModel.yAlign);
              } else {
                tooltipEl.classList.add("no-transform");
              }

              // eslint-disable-next-line @typescript-eslint/no-explicit-any
              function getBody(bodyItem: any) {
                return bodyItem.lines;
              }

              // Set Text
              if (tooltipModel.body) {
                const titleLines = tooltipModel.title || [];
                const bodyLines = tooltipModel.body.map(getBody);
                if (!Number.isNaN(Number(bodyLines[0])))
                  ReactDOM.render(
                    <CustomToolTip
                      title={titleLines[0]}
                      value={bodyLines[0]}
                    />,
                    tooltipEl
                  );
              }

              // `this` will be the overall tooltip
              // eslint-disable-next-line @typescript-eslint/ban-ts-comment
              // @ts-ignore
              // eslint-disable-next-line no-underscore-dangle,react/no-this-in-sfc
              const position = this._chart.canvas.getBoundingClientRect();

              // Display, position, and set styles for font
              tooltipEl.style.opacity = "1";
              tooltipEl.style.position = "absolute";
              tooltipEl.style.left = `${
                position.left + window.pageXOffset + tooltipModel.caretX
              }px`;
              tooltipEl.style.top = `${
                position.top + window.pageYOffset + tooltipModel.caretY
              }px`;
              tooltipEl.style.transform = "translate(-49%, -116%)";
            }
          },
          layout: {
            padding: {
              left: 0,
              right: 0,
              top: 10,
              bottom: 10
            }
          },
          legend: {
            display: false
          },
          scales: {
            yAxes: [
              {
                gridLines: {
                  borderDash: [3, 7],
                  color: "#253859",
                  lineWidth: 2,
                  zeroLineColor: "rgba(0, 0, 0, 0)",
                  drawTicks: false
                },
                ticks: {
                  padding: 20,
                  fontFamily: `"proxima-nova", sans-serif`,
                  fontSize: 11,
                  fontStyle: "bold",
                  fontColor: "white",
                  callback(value, index, values) {
                    if (index === values.length - 1) {
                      return "";
                    }
                    if (index <= 7)
                      if (stepSize <= 100) return `$${value.toLocaleString()}`;
                    return formatNumber(
                      value > 1_000_000
                        ? Number(value)
                        : Math.round(Number(value))
                    )
                      .split("")
                      .join(String.fromCharCode(8202));
                  },
                  suggestedMin: min,
                  stepSize,
                  maxTicksLimit: 7.1,
                  suggestedMax: max
                }
              }
            ],
            xAxes: [
              {
                gridLines: {
                  drawTicks: false,
                  color: "#253859",
                  lineWidth: 1,
                  drawOnChartArea: false
                },
                ticks: {
                  fontColor: "white",
                  fontFamily: `"proxima-nova", sans-serif`,
                  fontSize: 11,
                  fontStyle: "bold",
                  callback(value): string | number {
                    return value
                      .toString()
                      .split("")
                      .join(String.fromCharCode(8202));
                  },
                  padding: 20
                }
              }
            ]
          },
          plugins: {
            filler: {
              propagate: false
            }
          }
        },
        data: {
          labels: ["", "6 MONTHS", "1 YEAR", "3 YEARS", "5 YEARS", "10+ YEARS"],
          datasets: [
            {
              label: "Min",
              radius: 0,
              backgroundColor: "rgba(15, 212, 255, .15)",
              borderColor: "rgba(15, 212, 255, .15)",
              borderWidth: 0,
              fill: false,
              data: data.map(cv => cv.lowerBar),
              borderDash: [0, 1]
            },
            {
              fill: false,
              pointRadius: 6,
              data: data.map(cv => cv.nav),
              borderColor: "rgba(15,212,255,.3)",
              borderWidth: 2,
              pointBorderColor: "#03132F",
              pointBackgroundColor: "#1397B4",
              pointBorderWidth: 3
            },
            {
              label: "Max",
              backgroundColor: "rgba(15, 212, 255, .15)",
              borderColor: "rgba(15, 212, 255, .15)",
              fill: "-2",
              data: data.map(cv => cv.upperBar),
              radius: 0,
              borderWidth: 0,
              borderDash: [0, 1]
            }
          ]
        }
      });
      setChart(lineChart);
    }
  }, [chartRef, setChart, wealthProjections, nav]);
  return (
    <StyledRootCard label="portfolio growth" tooltipContent={tooltipContent}>
      <ChartCard>
        <canvas ref={chartRef} id="portfolio-growth" height="100" width="258" />
      </ChartCard>
    </StyledRootCard>
  );
};

export default PortfolioGrowthChart;
