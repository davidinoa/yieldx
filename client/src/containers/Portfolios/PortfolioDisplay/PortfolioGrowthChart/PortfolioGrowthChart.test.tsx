/* eslint-disable react/jsx-props-no-spreading */
import React from "react";
import { render } from "test/utils";
import PortfolioGrowthChart from "./PortfolioGrowthChart";

describe("PortfolioGrowthChart", () => {
  const props = {
    nav: 55_000,
    tooltipContent: "",
    wealthProjections: [
      {
        date: "2020-10-03",
        percentiles: [
          { percentile: 0.3, nav: 54946.92075798862 },
          { percentile: 0.5, nav: 56091.680667803754 },
          { percentile: 0.7, nav: 57236.44057761889 }
        ]
      },
      {
        date: "2021-01-03",
        percentiles: [
          { percentile: 0.3, nav: 55111.62300107334 },
          { percentile: 0.5, nav: 56749.54548375409 },
          { percentile: 0.7, nav: 58387.46796643484 }
        ]
      },
      {
        date: "2021-07-03",
        percentiles: [
          { percentile: 0.3, nav: 55717.487256860855 },
          { percentile: 0.5, nav: 58088.512683415094 },
          { percentile: 0.7, nav: 60459.53810996933 }
        ]
      },
      {
        date: "2022-07-03",
        percentiles: [
          { percentile: 0.3, nav: 57348.735602428416 },
          { percentile: 0.5, nav: 60861.968555671396 },
          { percentile: 0.7, nav: 64375.201508914375 }
        ]
      },
      {
        date: "2023-07-03",
        percentiles: [
          { percentile: 0.3, nav: 59259.59065845369 },
          { percentile: 0.5, nav: 63767.84402553859 },
          { percentile: 0.7, nav: 68276.09739262349 }
        ]
      },
      {
        date: "2025-07-03",
        percentiles: [
          { percentile: 0.3, nav: 63613.27936328368 },
          { percentile: 0.5, nav: 70002.44529976406 },
          { percentile: 0.7, nav: 76391.61123624444 }
        ]
      },
      {
        date: "2027-07-03",
        percentiles: [
          { percentile: 0.3, nav: 68547.72201991704 },
          { percentile: 0.5, nav: 76846.60541422582 },
          { percentile: 0.7, nav: 85145.4888085346 }
        ]
      },
      {
        date: "2030-07-03",
        percentiles: [
          { percentile: 0.3, nav: 76978.96975720959 },
          { percentile: 0.5, nav: 88387.71435828974 },
          { percentile: 0.7, nav: 99796.45895936989 }
        ]
      }
    ]
  };
  test("Renders correct line chart", () => {
    const { asFragment } = render(<PortfolioGrowthChart {...props} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
