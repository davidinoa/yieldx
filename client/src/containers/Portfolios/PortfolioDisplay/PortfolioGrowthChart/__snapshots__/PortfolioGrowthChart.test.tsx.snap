// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PortfolioGrowthChart Renders correct line chart 1`] = `
<DocumentFragment>
  .c0 {
  width: 100%;
  height: 100%;
  background: var(--dark);
  border: 1px solid var(--border-new);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: stretch;
  -webkit-justify-content: stretch;
  -ms-flex-pack: stretch;
  justify-content: stretch;
  border-radius: 0.5rem;
  padding: 1rem 1.15rem;
  min-height: 6rem;
}

.c0 > * {
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
}

.c1 {
  font-weight: bold;
  line-height: normal;
  -webkit-letter-spacing: 0.2em;
  -moz-letter-spacing: 0.2em;
  -ms-letter-spacing: 0.2em;
  letter-spacing: 0.2em;
  text-transform: uppercase;
  color: var(--lightest);
  position: relative;
  font-size: 0.65rem;
  padding: 0.25rem 0;
  margin-bottom: 0.25rem;
  max-height: 2rem;
}

.c1 > svg {
  margin: -0.5rem 0 0 0.35rem;
  vertical-align: middle;
  width: 0.85rem;
  height: 0.85rem;
}

.c2 {
  padding: 1.5rem 0.5rem 0;
}

@media (min-width:1600px) {
  .c0 {
    border-radius: 0.625rem;
    padding: 1rem 1.65rem 1.65rem 1.65rem;
  }
}

@media (min-width:1600px) {
  .c1 {
    font-size: 0.75rem;
    margin-bottom: 0.5rem;
  }
}

@media (min-width:1600px) {
  .c1 > svg {
    width: 1rem;
    height: 1rem;
  }
}

<div
    class="mb-1 col"
    data-testid="Portfolio-Metric-Card"
  >
    <div
      class="c0"
    >
      <div
        class="c1 d-flex align-items-center"
      >
        portfolio growth
      </div>
      <div
        class="c2 col"
      >
        <canvas
          class="chartjs-render-monitor"
          height="0"
          id="portfolio-growth"
          style="display: block; width: 0px; height: 0px;"
          width="0"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
